spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/ReactiveRedisStreamMessageProducer.java

```diff
@@ -0,0 +1,207 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.redis.inbound;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+
+import org.springframework.data.redis.connection.ReactiveRedisConnectionFactory;
+import org.springframework.data.redis.connection.stream.Consumer;
+import org.springframework.data.redis.connection.stream.ReadOffset;
+import org.springframework.data.redis.connection.stream.StreamOffset;
+import org.springframework.data.redis.core.StreamOperations;
+import org.springframework.data.redis.serializer.StringRedisSerializer;
+import org.springframework.data.redis.stream.StreamReceiver;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.redis.support.RedisHeaders;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import reactor.core.publisher.Flux;
+
+/**
+ * A {@link MessageProducerSupport} for Redis that reads message from a Redis Stream and publishes it to the provided
+ * output channel.
+ * By default this adapter reads message as a standalone client {@code XREAD} Redis command but we can switch to a
+ * Consumer Group feature {@code XREADGROUP} by setting {@link #consumerGroupName} and {@link #consumerName} fields.
+ *
+ * @author Attoumane Ahamadi
+ *
+ * @since 5.4
+ */
+public class ReactiveRedisStreamMessageProducer extends MessageProducerSupport {
+
+	private final ReactiveRedisConnectionFactory reactiveConnectionFactory;
+
+	private final Expression streamKeyExpression;
+
+	private EvaluationContext evaluationContext;
+
+	private StreamReceiver.StreamReceiverOptions<String, ?> streamReceiverOptions = StreamReceiver
+			.StreamReceiverOptions.builder().pollTimeout(Duration.ZERO).build();
+
+	private StreamReceiver<String, ?> streamReceiver;
+
+	private ReadOffset readOffset = ReadOffset.lastConsumed();
+
+	private boolean extractPayload = true;
+
+	private String streamKey;
+
+	@Nullable
+	private String consumerGroupName;
+
+	@Nullable
+	private String consumerName;
+
+	private boolean createConsumerGroupIfNotExist;
+
+	public ReactiveRedisStreamMessageProducer(ReactiveRedisConnectionFactory reactiveConnectionFactory,
+			String streamKey) {
+		this(reactiveConnectionFactory, new LiteralExpression(streamKey));
+	}
+
+	public ReactiveRedisStreamMessageProducer(ReactiveRedisConnectionFactory connectionFactory,
+			Expression streamKeyExpression) {
```

Comment:
There is no reason in such an expression complexity. This component is the beginning of the flow, so there is no special context to really evaluate a stream key on demand.

Reply:
I think I need a little briefing on whether or not to use `Expression`. Why and when to use `Expression` instead of `String`? In the outbound side I had to replace `String streamKey` by `Expression streamKeyExpression`. Why not here? 
Thanks

Reply:
We deal with expressions in the place where we have a dynamic context of some object, so the target value is resolved against that root object. On the outbound side it is a request message, therefore `streamKey` can be resolved at runtime every request we send to that outbound channel adapter. It also makes sense sometimes in polling inbound channel adapters when `receive()` is called by poller, so, when some environment in the application is changed in between polls, we get an actual value evaluating expression.

But if such an expression is going to evaluated only once during application startup, it does not give us any advantages to calculate the target value from the expression. Well, you still can use a SpEL, but as a matter of configuration phase - the target value will be populated to the inbound channel adapter only once anyway.